### PR TITLE
issue147/gpl 726 whats of value in this lh box then

### LINKED ISSUE #35: GPL-726 Whats of value in this LH box then?
This represents the lighthouse-ui work required for https://api.github.com/repos/sanger/lighthouse/issues/147

Original issue:
**User story**
GPL-726 | As Process Improvement I would like a LIMS feature for the new cherrypicking process where Sample Management staff can scan a Lighthouse Lab box and see how many plates are in the box, as well as how many plates have metadata, in order to prevent lab staff starting to process plates that are not ready to be cherrypicked.

**Who are the primary contacts for this story**
Scott T (R&D)
Lesley S (R&D)
Danni W
Ian J

**Acceptance criteria**
To be considered successful the solution must allow:

- [x] Do a quick mock up for Scott T (Proposer) & Danni W (User)
- [x] read 2D barcodes of Lighthouse boxes
- [x] list the plates in the box, indication plates with plate map data or not with tick/cross (maybe color signpost also green/red)
- [x] plates with plate map data to top panel, those with no plate map data to bottom
- [x] for each plate show the number of positives on the plate (sort plates by occupancy?) *1 
- [x] display summary below panel eg  'Total of 80 plates in box; 75 plates with plates maps, 5 without'
- [ ] Handle following error gracefully; a) Box barcode unknown b) Box empty
- [x] Be able to scan the contents of a plate individually (sometimes the user would want to check plates scanning them one by one to take the decision by plate)

*1 It maybe more *truthful* to display the number +ves that can be picked (eg not high Ct or Cambridge controls) however if this introduces significantly more work just do the positives 

**Dependencies**
This story is blocked by the following dependencies:
- #<issue_no.>
- sanger/<repo>#<issue_no.>

**Additional context**
Assumption this woudl be new UI in http://lighthouse-ui.psd.sanger.ac.uk/, called the 'Box buster'?

This story deprecates https://github.com/sanger/lighthouse-ui/issues/31

### PR DESCRIPTION
**Pull request by JamesGlover**
Add BoxBuster UI interface

fixes sanger/lighthouse#l147
fixes #35 

- Initial translation of wireframe to Vue
- Fix navbar collapse
- Add functiona box-buster
- Add polish and fix inconsistencies
- Improve feedback when falling back to plate lookup.
- Linter - Fix yarn lint issues

### CODE REVIEW

**Comment by JamesGlover** [comment]
UI component for: https://github.com/sanger/lighthouse/issues/147

File: pages/box_buster.vue [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+      </template>
+    </b-table>
+  </b-container>
+</template>
+
+<script>
+import lighthouse from '../modules/lighthouse_service'
+import { getPlatesFromBoxBarcodes } from '@/modules/labwhere'
+
+// import lighthouse from '../modules/lighthouse_service'
```
Comment (stevieing) [inline pages/box_buster.vue:51]: I guess you can remove this comment?

File: pages/box_buster.vue [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+const countWithMap = (accumulator, plate) =>
```
Comment (stevieing) [inline pages/box_buster.vue:52]: This is all good but wonder if it would sit better in a module especially as it is outside the actual component. I know they are simple but are there any tests for it? I just wasn't clear on exactly what is going on.

File: pages/box_buster.vue [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+  accumulator + (plate.plate_map ? 1 : 0)
+const countWithoutMap = (accumulator, plate) =>
+  accumulator + (plate.plate_map ? 0 : 1)
+const sumPositives = (accumulator, plate) =>
+  accumulator + plate.number_of_positives
+const mapFormatter = (value) => (value ? 'Yes' : 'No')
+const countFormatter = (value, _key, item) => (item.plate_map ? value : 'N/A')
+const extractError = (response) => {
```
Comment (stevieing) [inline pages/box_buster.vue:60]: I'm not really clear on what this does. Is there some edge case where an error is not returned? Did you see something in the tests that wasn't making sense?

File: pages/box_buster.vue [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+      if (this.barcode === '') return
+
+      const response = await getPlatesFromBoxBarcodes(this.barcode)
+      this.labwhereResponse = response
+      if (response.success) {
+        this.findPlates(response)
+      } else {
+        // If it isn't a box, perhaps its a plate.
+        // Requirements were that we should allow plate lookups
+        this.findPlates({ plateBarcodes: [this.barcode] })
```
Comment (stevieing) [inline pages/box_buster.vue:154]: You can use destructuring in the api call so you don't have to send an object or turn it into an array

File: modules/lighthouse_service.js [hunk 5]
```diff
@@ -31,6 +31,20 @@ const createPlatesFromBarcodes = async (moduleOptions) => {
   return responses
 }
 
+// Accepts a list of plateBarcodes in the moduleOptions
+// Send a GET request to the Lighthouse service API
```
Comment (stevieing) [inline modules/lighthouse_service.js:35]: Sharing my recently improved javascript knowledge here but if you do 

`const findPlatesFromBarcodes = async ({ barcodes }) => {
`
then you can do:

`params: { barcodes }`

Just saves you having to explicity name the variable as JavaScript infers it.

File: pages/box_buster.vue [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+    labwhereFeedback() {
+      const error = extractError(this.labwhereResponse)
+      return `${error}. Looking up barcode as plate.`
+    },
+    showTable() {
+      // We show the table if we've made a labwhere request, or have
+      // populates the plates via some other means.
+      return this.labwhereState !== null || this.plates.length !== 0
+    },
+    lighthouseFeedback() {
```
Comment (stevieing) [inline pages/box_buster.vue:125]: If you look in the index.vue there is a way of using states which seems quite cool for showing comments etc. So for example you can set the status to busy when a call is in progress and error when an error has been returned. I am using it in the printing and plan to extract it into a module and component of it's own.

File: test/modules/lighthouse_service.spec.js [hunk 7]
```diff
@@ -174,6 +174,72 @@ describe('lighthouse_service api', () => {
...
+      expect(mock).toHaveBeenCalledTimes(1)
+      expect(mock).toHaveBeenNthCalledWith(
+        1,
+        `${config.privateRuntimeConfig.lighthouseBaseURL}/plates`,
+        { params: { barcodes: plateBarcodes } }
+      )
+    })
+
+    it('#for multiple barcodes on success', async () => {
+      const plateBarcodes = ['aBarcode1', 'aBarcode2']
```
Comment (stevieing) [inline test/modules/lighthouse_service.spec.js:211]: Again here if you do:

`const barcodes = ['aBarcode1', 'aBarcode2']`

then you can do:

 
```
 const result = await lighthouse.findPlatesFromBarcodes({
        barcodes
      })
```

and:

```
expect(mock).toHaveBeenNthCalledWith(
        1,
        `${config.privateRuntimeConfig.lighthouseBaseURL}/plates`,
        { params: { barcodes } }
      )
```

File: test/pages/box_buster.spec.js [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+    const wrapper = mount(BoxBuster, { localVue })
+    await wrapper.setData({ labwhereResponse: { success: true } })
+    const header = wrapper.find('table').findAll('th')
+    expect(header.at(0).text()).toContain('Barcode')
+    expect(header.at(1).text()).toContain('Plate Map')
+    expect(header.at(2).text()).toContain('Number Of Positives')
+  })
+
+  it("only shows the table once we've scanned the Box", async () => {
+    const wrapper = mount(BoxBuster, { localVue })
```
Comment (stevieing) [inline test/pages/box_buster.spec.js:88]: From what I remember setdata will be deprecated in the future so you would add the data when you mount but was wondering why you are having to set the data afterwards and do an await? From what I see setting the data is not async? Maybe I am missing something.

**Review by stevieing — APPROVED** [review]
I have added a fair few comments but mainly just JavaScripty stuff that I have been looking at and a few bits about async. With regard to the comments about variable naming and destructuring I tried those out locally and they worked.

File: pages/box_buster.vue (hunk 3 above)
Comment (JamesGlover) [reply pages/box_buster.vue:60]: getPlatesFromBoxBarcodes returns both strings and Error objects as the error value of the attribute. Vue starts throwing up warnings if you pass an error object directly into a prop that is expecting a string. I can update the behaviour of getPlatesFromBoxBarcodes, but was away that was used in a fe other places.

Comment (JamesGlover) [reply pages/box_buster.vue:60]: Some of the additional handling was just to avoid the risk of silent errors, but is probably over-cautious.

File: pages/box_buster.vue (hunk 4 above)
Comment (JamesGlover) [reply pages/box_buster.vue:154]: I'm sorry, I'm not quite sure what you mean here.

File: modules/lighthouse_service.js (hunk 5 above)
Comment (JamesGlover) [reply modules/lighthouse_service.js:35]: That would mean I needed to do the inverse with the response from getPlatesFromBoxBarcodes though, at the moment I can just pass the response straight through.

File: pages/box_buster.vue (hunk 2 above)
Comment (JamesGlover) [reply pages/box_buster.vue:52]: These are all tested as part of the behaviour of the component itself, following the vue-test-utils guidelines of:

> You want to write maintainable tests for your Vue components. As a part of this goal, you want your tests to avoid including implementation details of your components. You'd rather focus on making your tests give you the confidence for which they are intended.

Obviously this wouldn't apply any more if I pulled them into a module, as they'd be exposed independently, so I'd add tests then.

File: pages/box_buster.vue (hunk 3 above)
Comment (stevieing) [reply pages/box_buster.vue:60]: Ah ok. If you look in index.vue we handle it slightly differently. So we have an alert message rather than an error message or success message. Then we have alerts each of which shows dependent on the status. Then if the error is an object it doesn't matter it just shows it.

Comment (JamesGlover) [reply pages/box_buster.vue:60]: Yeah, just remembered another issue about using the form feedback parameters, is that it still validates them, even if they aren't used. (Hence the defaultResponse.error bit). So annoyingly we need to return valid content for the field, even if there is no error.

File: test/pages/box_buster.spec.js (hunk 8 above)
Comment (JamesGlover) [reply test/pages/box_buster.spec.js:88]: I think you're thinking of set props, setData isn't flagged as deprecated:
https://vue-test-utils.vuejs.org/api/wrapper/setdata.html
You'll also notice that it uses await in the example.

It used to be that vue-test-utils would try to use magic to avoid the need to await, but they removed that ability as it was very flaky. The await in this case isn't so much the data update I believe, but for the subsequent render update.
